lib/RGBot.js

```diff
@@ -294,83 +593,153 @@ const RGBot = class {
+        options['entityNames'] = options.targetName && [options.targetName] || [];
+        return this.findEntities(options).shift()?.result
+    }
+
+    /**
+     * Find the nearest entity matching the search criteria.
+     *
+     * @param {object} [options={}]
+     * @param {string[]} [options.entityNames=[]] List of targetNames to consider
+     * @param {boolean} [options.attackable=false] Whether the entity must be attackable. If true finds only mob and player entities.
+     * @param {boolean} [options.partialMatch=false] Consider entities whose username or name partially match one of the targetNames
+     * @param {number} [options.maxDistance=null]  Max range to consider
+     * @param {number} [options.maxCount=1]  Max count of matching entities to consider
+     *
+     * @param {function(health:number,defense:number,distance:number):number} [sortValueFunction] Function of that should return the best target with the lowest value, so it sorts to front.
+     *
+     * @return {FindResult[]}
+     *
+     * By default, this sorts by distance, but a good example function for considering health would be { return -1 * ((20-health)-(distance*1.1)) }
+     *
+     * To get the 'best' entity result, call findEntities(...).shift().  Note that the result may be null|undefined if no entities were found
+     */
+    findEntities(options = {}, sortValueFunction = (health, defense, distance) => { return distance } ) {
         const attackable = options.attackable || false;
-        this.#log(`Searching for Entity ${targetName}`);
-        return this.#bot.nearestEntity(entity => {
-            if(entity.isValid && (!targetName || this.entityNamesMatch(targetName, entity) || entity.username === targetName)) {
+        const partialMatch = options.partialMatch || false;
+        const maxCount = options.maxCount || 1;
+        const maxDistance = options.maxDistance || null;
+        const entityNames = options.entityNames || [];
+        /**
+         * @type {Entity[]}
+         */
+        const potentialTargets = []
+
+        /**
+         *
+         * @type {Object.<number,number>}
+         */
+        const distances = {}
+
+        this.#bot.nearestEntity(entity => {
+            if(entity.isValid && (!entityNames || entityNames.length === 0 || entityNames.find(targetName => (entity.username && entity.username === targetName) || targetName === entity.name || (partialMatch && ( (entity.username && entity.username.includes(targetName)) || entity.name.includes(targetName)))))) {
                 if(!attackable || (attackable && (entity.type === 'mob' || entity.type === 'player'))) {
-                    return true;
+                    const distance = this.position().distanceTo(entity.position)
+                    if (distance>maxDistance) {
+                        return false // we ran out of 'close' entities
+                    }
+                    potentialTargets.push(entity)
```

Comment:
I wonder if this returns in increasing order? https://github.com/PrismarineJS/mineflayer/blob/master/docs/api.md#botentities

I just found it weird that we are using a function that returns the nearest entity, but then ignoring that entity, and then just utilizing the fact that it iterates through entities for us. I just took a look at the implementation for nearestEntity, and it's internally just calling `bot.entities`: https://github.com/PrismarineJS/mineflayer/blob/7fb0a83fdc11246cb170caf8c0030ba13ff4159c/lib/plugins/entities.js#L74-L91

Actually, looking at their implementation, we are calculating distances twice, since they do it in their function as well. Additionally, nearestEntity doesn't go through the bots in increasing distance order I believe, according to their implementation :/ all nearestEntity does is iterate through `bot.entities`, performs a distance calculation, and then additionally filters on the function that you provide.

My recommendation is to go with 
```
this.#bot.entities.forEach(your existing block of code)
```

Reply:
Just don't forget to filter out the entity that is the bot as well!